File: lib/domain/cultivation_plan/policies/cultivation_plan_optimization_complete_policy.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Policies
      module CultivationPlanOptimizationCompletePolicy
        module_function

        # @param plan_status [String]
        # @param field_cultivation_statuses [Array<String>]
        def should_mark_plan_completed?(plan_status:, field_cultivation_statuses:)
          return false unless plan_status.to_s == "optimizing"
          return false if field_cultivation_statuses.empty?

          field_cultivation_statuses.all? { |s| s.to_s == "completed" }
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/policies/cultivation_plan_optimization_complete_policy.rs
//! Ruby: `Domain::CultivationPlan::Policies::CultivationPlanOptimizationCompletePolicy`

/// True when agrr allocate returned at least one field schedule to persist.
pub fn allocation_has_field_schedules(allocation_result: &serde_json::Value) -> bool {
    allocation_result
        .get("field_schedules")
        .and_then(|v| v.as_array())
        .is_some_and(|schedules| !schedules.is_empty())
}

pub fn should_mark_plan_completed(plan_status: &str, field_cultivation_statuses: &[String]) -> bool {
    if plan_status != "optimizing" {
        return false;
    }
    if field_cultivation_statuses.is_empty() {
        return false;
    }
    field_cultivation_statuses
        .iter()
        .all(|s| s == "completed")
}

#[cfg(test)]
mod policies_cultivation_plan_optimization_complete_policy_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/cultivation_plan/policies_cultivation_plan_optimization_complete_policy_test.rs"));
}
